src/linux/functions.cpp

```diff
@@ -14,28 +14,87 @@
  * limitations under the License.
  */
 
+#include <sys/stat.h>
 #include <sys/mount.h>
+#include <mntent.h>
 #include <errno.h>
 #include "../mountutils.hpp"
 
 NAN_METHOD(UnmountDisk) {
   if (!info[1]->IsFunction()) {
-    return Nan::ThrowError("Callback must be a function");
+    return Nan::ThrowTypeError("Callback must be a function");
   }
 
   v8::Local<v8::Function> callback = info[1].As<v8::Function>();
 
   if (!info[0]->IsString()) {
-    return Nan::ThrowError("Device argument must be a string");
+    return Nan::ThrowTypeError("Device argument must be a string");
   }
 
   v8::String::Utf8Value device(info[0]->ToString());
-  int code = umount(reinterpret_cast<char *>(*device));
 
-  if (code == 0) {
-    YIELD_NOTHING(callback);
-  } else {
-    char *error_string = strerror(code);
-    YIELD_ERROR(callback, error_string);
+  const char *device_path = reinterpret_cast<char *>(*device);
+  const char *mount_path = NULL;
+
+  // Stat the device to make sure it exists
+  struct stat stats;
+
+  if (stat(device_path, &stats) != 0) {
+    v8::Local<v8::Value> argv[1] = {
+      Nan::ErrnoException(errno, "stat", NULL, device_path)
+    };
+    Nan::MakeCallback(Nan::GetCurrentContext()->Global(), callback, 1, argv);
+    return;
+  } else if (S_ISDIR(stats.st_mode)) {
```

Comment:
I don't understand this thread very much. In this particular case, checking for `S_ISDIR` and checking for `! S_ISBLK` would lead to the same effect (unmounting `/dev` would fail). However, this module is only meant to work in block devices, so `! S_ISBLK` would have the added benefit that trying to unmount a random file would file.

@jhermsmeier What does `umount2()` complains about when unmounting a simple file?